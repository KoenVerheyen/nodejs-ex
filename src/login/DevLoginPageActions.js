// @flow

export const SELECT_TOKEN = 'SELECT_TOKEN'
export const USER_INFO_RECEIVED = 'USER_INFO_RECEIVED'
export const RETRIEVE_USER_INFO_FAILED = 'RETRIEVE_USER_INFO_FAILED'
export const SELECT_IDENTITY = 'SELECT_IDENTITY'
export const SELECT_IDENTITY_SUCCESS = 'SELECT_IDENTITY_SUCCESS'
export const SELECT_IDENTITY_FAILED = 'SELECT_IDENTITY_FAILED'



type SelectTokenAction = {
    type: 'SELECT_TOKEN',
    token: number
}

type SelectIdentitySuccessfulAction = {
    type: 'SELECT_IDENTITY_SUCCESS',
    user: Object
}

type SelectIdentityFailedAction = {
    type: 'SELECT_IDENTITY_FAILED',
    error: Object
}

type RetrieveUserInfoFailedAction = {
    type: 'RETRIEVE_USER_INFO_FAILED',
    error: Object
}

type UserInfoReceivedAction = {
    type: 'USER_INFO_RECEIVED',
    user: Object
}

type SelectIdentityAction = {
    type: 'SELECT_IDENTITY',
    token: number,
    person: number,
    function: number,
    profile: number
}

export type DevLoginAction =
    | SelectTokenAction
    | UserInfoReceivedAction
    | RetrieveUserInfoFailedAction
    | SelectIdentityAction
    | SelectIdentitySuccessfulAction
    | SelectIdentityFailedAction

export function selectToken(token: number) : SelectTokenAction {
    return {
        type: SELECT_TOKEN,
        token: token
    }
}

export function userInfoReceived(user: Object): UserInfoReceivedAction {
    return {
        type: USER_INFO_RECEIVED,
        user: user
    }
}

export function retrieveUserInfoFailed(error: Object): RetrieveUserInfoFailedAction {
    return {
        type: RETRIEVE_USER_INFO_FAILED,
        error: error
    }
}




export function selectIdentity(token: number, identity: { person: number, function: number, profile: number}): SelectIdentityAction {
    return {
        type: SELECT_IDENTITY,
        token: token,
        person: identity.person,
        function: identity.function,
        profile: identity.profile,
    }
}

export function selectIdentitySuccessful(user: Object): SelectIdentitySuccessfulAction {
    return {
        type: SELECT_IDENTITY_SUCCESS,
        user: user
    }
}

export function selectIdentityFailed(error: Object): SelectIdentityFailedAction {
    return {
        type: SELECT_IDENTITY_FAILED,
        error: error
    }
}