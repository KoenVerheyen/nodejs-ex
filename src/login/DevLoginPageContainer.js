// @flow
import DevLoginPageComponent from './DevLoginPageComponent'
import {connect} from 'react-redux'
import {selectIdentity, selectToken} from './DevLoginPageActions'


let mapStateToProps = (state: Function, ownProps: Object): Object => {
    return {
        token: state.rad.iam.token,
        identities: state.rad.iam.user ? state.rad.iam.user.identities : [],
        loading: state.rad.iam.loading
    }
}

let mapDispatchToProps = (dispatch: Function, ownProps: Object): Object => {
    return {
        onSelectToken: (token: number) => {dispatch(selectToken(token))},
        postSelectIdentity: (token: number, identity: Object) => {dispatch(selectIdentity(token, identity))}
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(DevLoginPageComponent)