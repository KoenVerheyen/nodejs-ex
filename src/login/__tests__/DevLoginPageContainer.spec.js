import React from 'react'
import {shallow} from 'enzyme'
import configureMockStore from 'redux-mock-store'
import DevLoginPageContainer from '../DevLoginPageContainer'
import {selectIdentity, selectToken} from "../DevLoginPageActions";


const mockStore = configureMockStore([])

describe('DevLoginPageContainer', () => {

    test('store to component mapping without user', () => {
        let store = mockStore({
            rad: {
                iam: {
                    loading: false,
                    token: 'lrc',
                }
            }
        })

        let wrapper = shallow(<DevLoginPageContainer store={store} />)

        expect(wrapper.length).toEqual(1)
        expect(wrapper).toMatchSnapshot()
        let component = wrapper.find('DevLoginPageComponent')
        expect(component.length).toEqual(1)
        expect(component.props().token).toEqual('lrc')
        expect(component.props().loading).toEqual(false)

        component.props().onSelectToken('lrc')
        expect(store.getActions().length).toEqual(1)
        expect(store.getActions()[0]).toMatchObject(selectToken('lrc'))
    })

    test('store to component mapping with user', () => {
        let identities = [
            {
                person: {
                    id: 1,
                    lastName: 'Doe',
                    firstName: 'John'
                },
                function: {
                    id: 1,
                    name: 'functionLF'
                },
                profile: {
                    id: 1,
                    name: 'local rad coordinator'
                }
            },
            {
                person: {
                    id: 1,
                    lastName: 'Doe',
                    firstName: 'John'
                },
                function: {
                    id: 2,
                    name: 'functionEB'
                },
                profile: {
                    id: 1,
                    name: 'local rad coordinator'
                }
            }
        ]
        let store = mockStore({
            rad: {
                iam: {
                    loading: true,
                    token: 'lrc',
                    user: {
                        identities
                    }
                }
            }
        })

        let wrapper = shallow(<DevLoginPageContainer store={store} />)

        expect(wrapper.length).toEqual(1)
        expect(wrapper).toMatchSnapshot()
        let component = wrapper.find('DevLoginPageComponent')
        expect(component.length).toEqual(1)
        expect(component.props().token).toEqual('lrc')
        expect(component.props().loading).toEqual(true)
        expect(component.props().identities).toMatchObject(identities)
        expect(component.props().identities.length).toEqual(2)

        component.props().postSelectIdentity('lrc', {person: 1, function: 1, profile: 1})
        expect(store.getActions().length).toEqual(1)
        expect(store.getActions()[0]).toMatchObject(selectIdentity('lrc', {person: 1, function: 1, profile: 1}))
    })
})