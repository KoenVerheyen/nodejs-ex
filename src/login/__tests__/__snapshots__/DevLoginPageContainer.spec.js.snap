// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DevLoginPageContainer store to component mapping with user 1`] = `
<DevLoginPageComponent
  identities={
    Array [
      Object {
        "function": Object {
          "id": 1,
          "name": "functionLF",
        },
        "person": Object {
          "firstName": "John",
          "id": 1,
          "lastName": "Doe",
        },
        "profile": Object {
          "id": 1,
          "name": "local rad coordinator",
        },
      },
      Object {
        "function": Object {
          "id": 2,
          "name": "functionEB",
        },
        "person": Object {
          "firstName": "John",
          "id": 1,
          "lastName": "Doe",
        },
        "profile": Object {
          "id": 1,
          "name": "local rad coordinator",
        },
      },
    ]
  }
  loading={true}
  onSelectToken={[Function]}
  postSelectIdentity={[Function]}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": [Function],
    }
  }
  token="lrc"
/>
`;

exports[`DevLoginPageContainer store to component mapping without user 1`] = `
<DevLoginPageComponent
  identities={Array []}
  loading={false}
  onSelectToken={[Function]}
  postSelectIdentity={[Function]}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": [Function],
    }
  }
  token="lrc"
/>
`;
